### ISSUE TITLE
Non-axis aligned static geometry is visible behind portals

### ISSUE THREAD

**Issue by mwpenny** [mwpenny/portal64-still-alive#73]
#### Version of the game 
Latest (f35bf3d2a2fbc62c08df614dc258da1ae8119d43)

#### Original Hardware or Emulator?
Emulator: Ares 134

#### What is the bug?
When a portal is fired, looking through the opposite portal can show non-axis aligned static level geometry which is located behind the first portal.

This can be seen in chamber 15 (#39). The problem does not occur with dynamic (i.e., animated) geometry such as in chamber 12.

#### Steps to reproduce the behavior:
1. In chamber 15, play until the large room before the hazard hallway
2. Fire first portal on the angled wall (normally used to redirect the energy ball)
3. Fire second portal on another wall and look through it (i.e., look out of angled wall portal)
4. Observe that geometry behind the angled wall shows through

#### Screenshots
Problem in chamber 15:
![portal 2024-05-30 23-42-21](https://github.com/mwpenny/portal64-still-alive/assets/9356790/d2987099-1f2b-4308-9d1d-afecc3108a56)

No problem in chamber 12:
![portal 2024-05-31 00-01-51](https://github.com/mwpenny/portal64-still-alive/assets/9356790/59dc2e05-a559-4f10-9820-5374d2602376)

**Comment by mwpenny** [comment]
I originally thought this was due to the view through the portal not being rotated properly. It turns out it's because the logic for culling static geometry only uses axis-aligned bounding boxes. So as a result, bounding boxes for rotated geometry can be larger than the geometry itself and clip through the back of portals to be considered visible. In other words, rotated surfaces aren't the problem - the rotated objects behind them are.

This is made worse by the fact that level BVHs can potentially group multiple static objects in their leaf nodes, meaning one object can make multiple visible.

In short, the issue is that coarse culling (which makes sense for performance reasons) is being used in a situation where it needs to be more granular (views through portals). Animated objects use rotated bounding boxes for more precise culling which is why the problem doesn't occur in chamber 12.

**Comment by mwpenny** [comment]
Added a new optional argument for [`@static`](https://github.com/mwpenny/portal64-still-alive/blob/master/documentation/levels/level_objects/static.md) level objects: `precise_culling`. Static geometry marked for precise culling will be culled based on its exact (i.e., rotated) bounding box, in the same way dynamic/animated geometry is.

Precise culling is more expensive than standard culling so it is opt-in. It is intended only for the few situations that can hit the edge cases described above. So far, chamber 15 (and I know it will be needed at the beginning of chamber 19 as well).

End result:
![portal 2024-07-26 18-52-12](https://github.com/user-attachments/assets/d9568621-3a57-4ddd-88e1-56c404458ee4)

**Comment by hackgrid** [comment]
Cool stuff, thank you, you are very thorough! :-)
